Allow histograms for timer metrics

bakins opened the pull request:
Based on comments in https://github.com/prometheus/statsd_exporter/issues/19 this allows the use of a YAML config file that allows setting histograms for timer metrics per mapping or as the default timer type. Previous behavior is the same when using the "old" config file format.

This is fairly rough and could probably be cleaner.  I did the minimum possible to enable histograms as an experiment.  Consider this enough to start a conversation.

bakins commented:
Anything I can do to help on this? We've been running my fork for a while and it seems to be working fine.

gerricom commented:
@bakins: Could you provide a sample configuration file? I do not understand how to map the labels with your yaml-solution... 😕

bakins commented:
@gerricom sorry for the delay.  

This PR was to start the discussion. I can add documentation if people think this is useful.

Example:

"normal" config:

```
test.dispatcher.*.*.*
name="dispatcher_events_total"
processor="$1"
action="$2"
outcome="$3"
job="test_dispatcher"
```

YAML version of this:
```yaml
mappings:
- match: test.dispatcher.*.*.*
  labels:
    name: "dispatcher_events_total"
    processor: "$1"
    action: "$2"
    outcome: "$3"
    job: "test_dispatcher"
```

To enable histograms for a single mapping:
```yaml
mappings:
- match: test.dispatcher.*.*.*
  timer_type: histogram
  labels:
    name: "dispatcher_events_total"
...
```

To enable histograms by default:
```yaml
defaults: 
  timer_type: histogram
```

gerricom commented:
@bakins No problem! Thanks for the example. I must have simply missed the match parameter in the YAML configuration.

So after playing around with it: This looks quite useful and I'd like to see this merged. Changing over to the YAML configuration style is very easily done and straightforward. Histograms seem to work as expected.

juliusv commented:
@bakins Sorry for ignoring this for so long. I think in general changing this to YML is a great idea, to get rid of yet another custom language. Taking a closer look now.

juliusv commented on mapper.go:
```diff
@@ -102,40 +113,90 @@ func (m *metricMapper) initFromString(fileContents string) error {
 	return nil
 }
 
+func (m *metricMapper) initFromYamlString(fileContents string) error {
```
style: Yaml -> YAML (https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

juliusv commented on mapper_test.go:
```diff
@@ -148,7 +148,7 @@ func TestMetricMapper(t *testing.T) {
+
 				test.foo
 				name="name_foo"
 				label="bar"`,
```
Would be great to have some YAML-based tests here of course.

juliusv commented on exporter.go:
```diff
@@ -263,11 +299,28 @@ func (b *Exporter) Listen(e <-chan Events) {
+					timerType = b.mapper.Defaults.TimerType
+				}
+
+				if timerType == "histogram" {
```
As far as I can see, the timer type string isn't validated at config load time, but probably should be?

And then I'd change this to a `switch timerType {...}` here with a default case that panics. Could be good to extract the strings into named constants too.

juliusv commented on exporter.go:
```diff
@@ -263,11 +299,28 @@ func (b *Exporter) Listen(e <-chan Events) {
+
+				if timerType == "histogram" {
+					histogram := b.Histograms.Get(
+						b.suffix(metricName, "timer"),
```
Hmm, I think we should only suffix metrics that don't have an explicitly mapped metric name (if there is a mapping, use exactly the metric name provided in the mapping). But I see we also do that for gauges and counters already. So that should probably be fixed in a separate change.

bakins commented:
This now validates the timer type when it reads the configuration. Also addresses a couple of the style issues. Let me know what you think. I'll add docs to this PR after we think the code looks good.

juliusv commented on mapper.go:
```diff
@@ -107,40 +118,92 @@ func (m *metricMapper) initFromString(fileContents string) error {
+func (m *metricMapper) initFromYAMLString(fileContents string) error {
+	var n metricMapper
+
+	fmt.Println(fileContents)
```
Remove debug Println here and Printf below.

juliusv commented on exporter.go:
```diff
@@ -273,17 +309,44 @@ func (b *Exporter) Listen(e <-chan Events) {
-					log.Errorf(regErrF, metricName, err)
-					conflictingEventStats.WithLabelValues("timer").Inc()
+				switch t {
+				case "histogram":
```
`"histogram"` -> `timerTypeHistogram`?

grobie commented:
Could you add some tests for the new feature?

bakins commented:
Added basic config tests.

drawks commented:
This looks like a very nice change. One critique of mine would be to separate the conversion to yaml config and the addition of histograms as two separate PRs. I'm a big fan of both changes though personally.

juliusv commented on mapper.go:
```diff
@@ -107,40 +118,105 @@ func (m *metricMapper) initFromString(fileContents string) error {
+			return fmt.Errorf("invalid match: %s", currentMapping.Match)
+		}
+
+		// check that label is correct
```
nit: start all comments upper-cased, end with period.

juliusv commented on mapper.go:
```diff
@@ -107,40 +118,105 @@ func (m *metricMapper) initFromString(fileContents string) error {
+		// check that label is correct
+		for k, v := range currentMapping.Labels {
+			label := fmt.Sprintf("%s=%q", k, v)
+			if len(labelLineRE.FindStringSubmatch(label)) != 3 {
```
It is a bit complicated to first make a k=v string out of this and then match the entire string against `labelLineRE`. The label value can be anything, so it would be better to just check whether the label name (`k`) matches `identifierRE`.

bakins replied:
Not sure why I did it this way. fixing.

juliusv commented on mapper.go:
```diff
@@ -107,40 +118,105 @@ func (m *metricMapper) initFromString(fileContents string) error {
+		}
+
+		if _, ok := currentMapping.Labels["name"]; !ok {
+			return fmt.Errorf("Line %d: metric mapping didn't set a metric name", i)
```
"Line" -> "line" (start error strings lower-cased)

bakins replied:
FWIW, errors in `updateMapping` in this same file use uppercase "L".  I'll change my to lower.

bakins commented:
@juliusv Addressed your comments.

@drawks I can split this up if that's the consensus.  YAML change would need to be merged first, IMO. In https://github.com/prometheus/statsd_exporter/issues/19 it seems people did not want to mangle the current config format any more.  I guess splitting or not is up to @juliusv ?

bakins commented:
Added initial docs.

juliusv commented:
@bakins I think it's fine to just merge it as it is this time (YAML + histograms combined), although in the future I would recommend separating changes into multiple PRs.

Thanks for all the patience and sorry for being so slow here! I'll merge it and then we can still do followup cleanups as needed.

bakins commented:
@juliusv thanks. As with most things, this PR started very simple...
